### ISSUE TITLE
Window manager decorations are overridden by Flutter desktop apps

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#68253]
I use a window manager with very specific window decorations (ratpoison) but applications built with Flutter don't function well with that setup: they get a title bar and buttons like minimize/maximize that make zero sense in that environment.

We should honor the window manager.

**Comment by robert-ancell** [comment]
GTK uses [a setting](https://developer.gnome.org/gtk3/stable/GtkSettings.html#GtkSettings--gtk-decoration-layout) to determine the buttons to show. Perhaps there is some code that should be added somewhere to detect if you are running in ratpoison and set this appropriately?

**Comment by robert-ancell** [comment]
The buttons are drawn by GTK, so there's no standard way to determine what functions the window manager supports. Old style X windows have the titlebar rendered by the WM so this allowed different WMs to decide what buttons to show.

In your case, the easiest thing might be to edit `my_application.cc`and remove the headerbar, i.e.
```c
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));
  //GtkHeaderBar *header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
  //gtk_widget_show(GTK_WIDGET(header_bar));
  //gtk_header_bar_set_title(header_bar, "test2");
  //gtk_header_bar_set_show_close_button(header_bar, TRUE);
  //gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  gtk_window_set_title(window, "test2");
  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));
```

**Comment by robert-ancell** [comment]
Digging a bit further, the appropriate solution is that ratpoison should set the `Gtk/DecorationLayout` [XSETTING](https://specifications.freedesktop.org/xsettings-spec/xsettings-latest.html) to `"close"` - this would ensure all GTK applications show the correct buttons when running inside that window manager. I had a quick look at the ratpoison source code and it doesn't appear to be using XSETTINGS currently.

**Comment by robert-ancell** [comment]
And if it's just annoying you, then you can edit `/home/bob/.config/gtk-3.0/settings.ini` and set the following to do this for your machine:
```
[Settings]
gtk-decoration-layout=close
```

**Comment by Hixie** [comment]
This seems like a bug in GTK then? Why would a WM need to know about GTK to get GTK applications to act correctly?

**Comment by Hixie** [comment]
(To be clear, I don't want _any_ decorations to be drawn by the app. That's what the window manager is for.)

**Comment by robert-ancell** [comment]
Window management has somewhat changed over the years. If you have a Wayland compositor you might have no server side decorations at all.

For this to work automatically WM needs to express the way it manages windows so the toolkit can adjust client side decorations as necessary. There's a [bug in GTK](https://gitlab.gnome.org/GNOME/gtk/-/issues/2206) about removing the shadows in that case. Headerbars are a bit more complex as they could contain controls essential to the application.

The only method I know of currently is to check the value of `$XDG_CURRENT_DESKTOP` in your app and decide if a headerbar should be used in that case.

**Comment by Hixie** [comment]
By default, we should work with ratpoison and other window managers that might not know about GTK. Currently we don't.

**Comment by robert-ancell** [comment]
I think I've found a reasonable compromise in https://github.com/flutter/flutter/pull/68452 that will allow the majority of users (i.e. Ubuntu users) have modern and consistent decorations and other users to have traditional behaviour.
